Track and write measurement time for area

gerritholl opened the pull request:
Track measurement times such that a representative measurement time can be estimated and written for an area.

To include valid time in file headers for supported writers, add `reader_kwargs: {"track_time": True}` to `Scene.__init__` and
`resample_coords=True` to `Scene.resample`.

When storing as `ninjogeotiff`, time coordinates will be averaged and stored in the `ninjo_ValidTimeID` tag.

An example is included in the PR under `doc/source/examples/valid_time.rst`.

Those two PRs must be merged first:

* #3178 
* #3371 

Next steps, either within this PR or within one or more later PRs:

- [ ] Add other readers
- [x] Add other writers
- [ ] Add a way to include the valid time in the filename (deferred)
- [x] Investigate reduce_data performance penalty (see [this comment](https://github.com/pytroll/satpy/pull/3168#issuecomment-4207260455))

The latter is difficult to do without triggering a computation, so doing this efficiently should probably be done as postprocessing beyond the scope of this PR.

```python
import os
from satpy import Scene
from glob import glob
from satpy.utils import debug_on; debug_on()
seviri_files = glob("/media/nas/x21308/scratch/SEVIRI/202103300900/H-000*")
sc = Scene(filenames={"seviri_l1b_hrit": seviri_files}, reader_kwargs={"track_time": True})
sc.load(["IR_108"])
ls = sc.resample("nqeuro3km", resample_coords=True, resampler="gradient_search")
ls.save_dataset(
        "IR_108",
        "{platform_name}-{sensor}-{name}-{area.area_id}-{start_time:%Y%m%d%H%M}-geotiff.tif",
        writer="ninjogeotiff",
        ChannelID="IR 10.8",
        DataType="GORN",
        PhysicValue="Temperature",
        PhysicUnit="K",
        SatelliteNameID="METEOSAT Europe-Atlantic",
        fill_value=0)
```

 - [x] Closes #3161 <!-- remove if there is no corresponding issue, which should only be the case for minor changes -->
 - [x] Tests added <!-- for all bug fixes or enhancements -->
 - [x] Fully documented <!-- remove if this change should not be visible to users, e.g., if it is an internal clean-up, or if this is part of a larger project that will be documented later -->
 - [x] Add your name to `AUTHORS.md` if not there already

gerritholl committed:
Start unit test for SEVIRI time tracking

Start unit test to track time for the SEVIRI HRIT reader.  This is to
prepare an implementation of
https://github.com/pytroll/satpy/issues/3161.  There is no
implementation yet.

gerritholl committed:
First skeleton on how implementation might look like

gerritholl committed:
First implementation for seviri per-pixel times

Added a first implementation for SEVIRI per-pixel times in ancillary
variables.  So far just takes the scanline time for all pixels in the
scanline.

gerritholl committed:
in a variable?

gerritholl commented:
Maybe it fits better in a coordinate variable, but not sure if it's valid to have those be not-unique.

gerritholl committed:
Add time as coordinate

Add time as a coordinate rather than as an auxiliary variable.  Start
working on resampling time coordinates, so far with a failing test.

gerritholl committed:
Resample time coordinates

gerritholl committed:
resample time coordinate for real

gerritholl committed:
Implement time-tracking as floats

Implement time tracking as a float coordinate for SEVIRI L1B HRIT.
Retain this through resampling.  Add documentation on how to use this.

gerritholl committed:
Add test for writing ValidDateID

gerritholl committed:
Write ValidDateID to ninjogeotiff header

gerritholl committed:
Adapt ninjo tag test

Adapt ninjo tag test to account for the novelty of optional dynamic
tags.

gerritholl committed:
Refactor getting valid time out of ninjogeotiff writer

Refactor getting the valid time out of the ninjogeotiff writer, so that
the functionality is available elsewhere.

pnuu committed:
Refactor Scene._resampled_scene()

pnuu committed:
Move source area access to _reduce_data()

gerritholl committed:
Test writing valid time to filename

pnuu committed:
Refactor data reduction

gerritholl committed:
Write valid time to filename

gerritholl committed:
Pass dynamic fileds as a set.

gerritholl committed:
Add documentation on writing valid time

gerritholl committed:
Merge branch 'refactor_resampled_scene' into feature-valid-time

Fixing a merge conflict.

gerritholl committed:
Remove unneeded code

Remove code leftover from an earlier implementation attempt, that is no
longer needed.

gerritholl committed:
Doc fixes

gerritholl committed:
clarify writer docs

gerritholl committed:
Improve test coverage

pnuu commented on satpy/readers/seviri_l1b_hrit.py:
```diff
@@ -453,6 +456,19 @@ class HRITMSGFileHandler(HRITFileHandler):
+                            reader='seviri_l1b_hrit',
+                            reader_kwargs={'track_time': True})
+
+    Those coordinates are kept in resampling iff passing the argument
```
Not sure if the "and only if" is even necessary, seems redundant to me.

pnuu commented on satpy/tests/writer_tests/test_ninjogeotiff.py:
```diff
@@ -985,3 +991,30 @@ def test_str_ids(test_image_small_arctic_P):
+        ChannelID="trollchannel",
+        DataType="GORN",
+        DataSource="dowsing rod")
+    src = rasterio.open(fn.replace("{valid_time:%Y%m%d%H%M%S}",
```
Should the file be closed?

pnuu commented on satpy/tests/writer_tests/test_utils.py:
```diff
@@ -0,0 +1,50 @@
+                        dims=("y", "x"),
+                        attrs={"units": "seconds since 2222-02-02T22:22:20"})})
+
+    assert get_valid_time(da) == datetime.datetime(2222, 2, 2, 22, 22, 22)
+
+    da = xr.DataArray(np.zeros((3, 3)), dims=("y", "x"), name="Liedik")
+    with pytest.raises(ValueError,
+                       match="Dataset Liedik has no time coordinate."):
+        get_valid_time(da)
+
+    da = xr.DataArray(np.zeros((3, 3)), dims=("y", "x"), name="Liedik", attrs={"name": "Gámasčearru"})
+    with pytest.raises(ValueError,
+                       match="Dataset Gámasčearru has no time coordinate."):
+        get_valid_time(da)
```
These should be separate test cases.

gerritholl replied:
Split in two in c3626c430, one for has valid time, one for has no valid time.  Or should it even be split in three?  Should each piece of behaviour being tested come with its own testing function, or is that exaggerated?

gerritholl replied:
Split in three in [d619238](https://github.com/pytroll/satpy/pull/3168/commits/d619238e9758916f233a79b87422f9097d2f8ef2).

pnuu commented on satpy/writers/core/utils.py:
```diff
@@ -0,0 +1,44 @@
+from xarray.coding.times import decode_cf_datetime
+
+
+def get_valid_time(dataset):
+    """Get the valid time for a dataset.
+
+    If a dataset has time coordinates, get the arithmetic mean time as a
+    representative time for the dataset.
```
I wouldn't name this function as "valid" time but "average" or "mean" (observation) time. Or maybe add a kwarg that controls the behaviour? Others might have different definition of "valid" time. The other options could just be `NotImplementedError` for now.

gerritholl replied:
Renamed `get_mean_time` in c3626c430.

pnuu commented on satpy/writers/ninjogeotiff.py:
```diff
@@ -418,12 +430,22 @@ def get_date_id(self):
+        That's seconds since UNIX Epoch for a representative time for the
+        image.
+        """
+        dt = get_valid_time(self.dataset)
```
`dt` is used for `import datetime as dt`, so this is very confusing.

gerritholl replied:
Fixed in c3626c430.

pnuu reviewed (changes requested):
Some comments and change requests inline.

gerritholl committed:
Changes based on Panus review

Pased on Panus review, renamed get_valid_time to get_mean_time.  The tag
in the ninjogeotiff writer is still called ValidDateID, pending a decision
by NinJo developers.  Also renamed some local variables to avoid confusion
with conventions for imported module shorthands.

gerritholl committed:
Merge branch 'main' into feature-valid-time

gerritholl committed:
valid time is now mean time

gerritholl committed:
valid time is now mean time, also in docs

gerritholl committed:
fix small error in doc index.rst

gerritholl committed:
More doc fixes re. valid/mean time

gerritholl committed:
Fix doc reference errors

gerritholl committed:
Split remaining ValueError test in two

gerritholl committed:
Merge branch 'main' into feature-valid-time

Fixing two merge conflicts

gerritholl committed:
Resampling time coordinate fails if reduce_data=False

Resampling a time coordinate fails if data are not reduced
(``reduce_data=False``).  This commit adds a test reproducing this
error, but does not implement a solution yet.

gerritholl committed:
Merge branch 'main' into feature-valid-time

gerritholl committed:
Add SingleBandCompositor time test and refactor

Add a test function for testing that a SingleBandCompositor retains the
time coordinate (test passes).  In the process, refactor the single band
compositor tests to be based on pytest rather than pyunit.

gerritholl committed:
Test time coordinate retained in GenericCompositor

Add a test to confirm that the time coordinate is retained upon calling
the `GenericCompositor`.  This test currently fails, because the time
coordinate is explicitly removed by match_data_arrays.  This removal was
introduced by https://github.com/pytroll/satpy/issues/795 because
non-equal time coordinates can cause problems in compositing later on.
A solution satisfying both use cases is still outsanding.

gerritholl committed:
Expand tests for time coordinate compositing

Add more tests covering the different scenarios of time coordinates in
compositing.

gerritholl committed:
Make sure time info in tests has epoch data

gerritholl committed:
Average time coordinates

When creating a composite where more than zero components have a time
coordinate, calculate the arithmetic mean of those time coordinates.

gerritholl committed:
Average the coords not data values

It should be the time coordinate averaged, not the data that have the
time coordinates.  Units should be retained, and this should be tested,
that they are retained.

gerritholl committed:
Fix non-reduced coordinate resampling bug

The time coordinate resampling was failing if reduce_data=False, because
reduce_data=True has a side-effect of adding the area information to the
attributes.  Add this explicitly for the time coordinate.

gerritholl committed:
Retain time coordinate upon sunz correction

gerritholl committed:
Retain time coordinates in more cases

Retain time coordinates in several more cases where they were
accidentally dropped, such as in filler compositors and
DayNightCompositor.

gerritholl committed:
Verify no early dask computes with times

Verify that the times do no lead to early dask computes

gerritholl committed:
Avoid early computation in time coordinate

gerritholl committed:
remove check_times from tests

gerritholl committed:
Avoid dropping G and B

The time coordinate was getting a bands=R dimension, because it was
taking the non-time coordinates from the first projectable.  This was
then leading to the G and B dimensions being removed when the time
coordinate was being assigned.

Add a test to verify that this is fixed.

gerritholl committed:
Avoid dropping G and B

The time coordinate was getting a bands=R dimension, because it was
taking the non-time coordinates from the first projectable.  This was
then leading to the G and B dimensions being removed when the time
coordinate was being assigned.

gerritholl committed:
Add test to ensure no early dask computations

Add a test to ensure we are not having early dask computations due to
filling with the time coordinate.  The test is currently failing.

gerritholl committed:
Combine times prior to compositing

Combine times prior to compositing, avoiding an accidental early dask
computation in fillna.

gerritholl committed:
Merge branch 'feature-valid-time' of github.com:gerritholl/satpy into feature-valid-time

gerritholl committed:
Repair case with 1-D time coordinate

gerritholl committed:
WIP toward better laziness

Improve laziness for time coordinate tracking.  Not there yet.

gerritholl committed:
Lazy tag calculations for mean time

Calculate the tag value for the mean time lazily.

Putting the mean time in the filename is not supported for now.  Will be
readded in a later commit, but with a warning that this triggers an
early computation and is thus not recommended if performance matters.

gerritholl commented:
Writing the valid time in the headers now works lazily.  For the filename, lazy evaluation is too hard and I will probably not implement this.  Users who need this should use a second step renaming the file based on information from the header.

gerritholl committed:
Move atmospheric modifier tests to own test module

Mode tests for modifiers in modifiers/atmosphere.py to their own test
module in tests/modifier_tests/test_atmosphere.py.

gerritholl committed:
Add unit test for CO2Corrector

gerritholl committed:
Merge branch 'test-co2-cerroctor' into feature-valid-time

gerritholl committed:
Add test to confirm CO2 corrector does not compute

Add a test to confirm that the CO2Corrector modifier does not trigger a
dask computation in the presence of non-identical time coordinates.

gerritholl committed:
Retain time coordinates without computation

Retain time coordinates without computation when applying the CO2
correction.

gerritholl committed:
Retain time coordinates in arithmetic compositors

Retain time coordinates in the arithmetic compositors and the CO2
correction, such that they are retained in the convection RGB.

gerritholl commented:
**EDIT:** This comment is out of date.  There is no longer a performance penalty.

Original comment:

---

This PR is strongly increasing the overhead of `reduce_data` when resampling, even if time tracking is not switched on.

Calling `Scene.resample`:

| branch | reduce_data | time tracking | [caching](https://pyresample.readthedocs.io/en/latest/howtos/configuration.html#cache-geometry-slices) | duration
| ------------- | ------------- | ---- | --- | --- | 
| main | False  | N/A | False | 0.97 
| main  | True | N/A | False | 8 
| main | False | N/A | True | 0.96
| main | True | N/A | True | 0.97
| 3168 | False | False | False | 0.96
| 3168 | False | False | True | 0.96
| 3168 | False | True | False | 1.11
| 3168 | False | True | True | 1.20
| 3168 | True | False | False | 42.7
| 3168 | True | False | True | 0.99
| 3168 | True | True | False | 83.4
| 3168 | True | True | True | 1.2

NB: This is just for calling resample — no computation.  See also https://github.com/pytroll/satpy/issues/2620.

I don't yet understand why the overhead for `resample` with `reduce_data=True` increases from 8 to 43 seconds even if no time tracking is actually used.  Needs to be investigated more.

gerritholl commented:
For ninjogeotiff, the NinJo team have agreed this should go in a tag called `ninjo_ValidTimeID` (not `ValidDateID`).

(In case a NinJo developer or someone with access to the internal NinJo JIRA reads this: details in NinJo JIRA-38947).

lahtinep committed:
Move class methods to functions

lahtinep committed:
Move unnecessary code out of try block

lahtinep committed:
Use dataclass to hold resampling context

lahtinep committed:
Merge branch 'main' into refactor_resampled_scene

gerritholl committed:
Merge branch 'main' into feature-valid-time

lahtinep committed:
Merge branch 'main' into refactor_resampled_scene

gerritholl committed:
Merge branch 'main' into feature-valid-time

with one merge conflict

gerritholl committed:
Rename to ValidTimeID and drop filename support

Rename the NinJo tag from ValidDateID to ValidTimeID as agreed by the
NinJo team.

Drop support for putting the mean time in the filename; this is not
currently possible to do without triggering an early computation.

gerritholl commented:
The worsening performance with `_reduce_data` when not using time tracking is due to https://github.com/pytroll/satpy/pull/3178.

The even-more-worsening performance with time tracking is because reduce data is called twice as often.

gerritholl committed:
Fix performance regression

Fix a performance regression in _reduce_data, where the reductions
dictionary was no longer used correctly, leading to reduce_data being
called for every dataset instead of every source area.

gerritholl commented:
Both problems are solved with https://github.com/pytroll/satpy/pull/3178/commits/f70d80f2c2d5a52a222a45920c36a0e36a7d64d1.  The superfluous calls to `_reduce_data` are not a problem any more (little overhead), because on the second run it finds the source areas in the reductions dictionary.  The overhead merely increases from 8.3 to 8.6 seconds (previously from 8.3 to 83.4 seconds).

gerritholl committed:
fix reduction caching

Fix caching when reduction, so reduce data doesn't reduce the same
source area twice.

gerritholl committed:
Merge branch 'refactor_resampled_scene' into feature-valid-time

fix major merge conflict, probably introducing errors

gerritholl committed:
Some more refactoring in _process_one_dataset

gerritholl committed:
Remove dynamic fields from filename

Dynamic fields in the filename do not work, they trigger an early
computation.  I had removed them in an earlier commit, but missed the
rollback in the CF and simple image writers.

gerritholl committed:
Refactor _verify_times and improve error messages

Refactor _verify_times with three small subfunctions, thus reducing the
cyclomatic complexity and hopefully satisfying codescene.  And while at
it, improve error messages for the IncompatibleTimes or ValueError
exceptions.

gerritholl committed:
Refactor fixtures in test_atmosphere.py

Refactor the dask/time fixtures in test_atmosphere.py to reduce code
duplication.

gerritholl committed:
Remove reference to disabled "valid time in filename" in doc
